src/Parallel/Callback.hpp

```diff
@@ -0,0 +1,95 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines Parallel::Callback.
+
+#pragma once
+
+#include <pup.h>
+#include <tuple>
+#include <utility>
+
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/Invoke.hpp"
+
+namespace Parallel {
+/// A class that holds a function that can be invoked at a later time.
+class Callback : public PUP::able {
+ public:
+  WRAPPED_PUPable_abstract(Callback);  // NOLINT
+  Callback() = default;
+  Callback(const Callback&) = default;
+  Callback& operator=(const Callback&) = default;
+  Callback(Callback&&) = default;
+  Callback& operator=(Callback&&) = default;
+  virtual ~Callback() = default;
+  virtual void invoke() noexcept = 0;
+};
+
+/// Wraps a call to a simple action and its arguments.
+template <typename SimpleAction, typename Proxy, typename... Args>
+class SimpleActionCallback : public Callback {
+ public:
+  WRAPPED_PUPable_decl_template(SimpleActionCallback);  // NOLINT
+  SimpleActionCallback(Proxy proxy, Args&&... args)
+      : proxy_(proxy), args_(std::make_tuple<Args...>(args...)) {}
+  void invoke() noexcept override {
+    std::apply(
+        [this](auto&&... args) noexcept {
+          Parallel::simple_action<SimpleAction>(proxy_, args...);
+        },
+        std::move(args_));
```

Comment:
the `std::move` of the args means that the callback can only be safely invoked once. On a quick inspection, I can't tell whether or not that is a feature of the charm version, but it might be good to mention that constraint in the dox.